dev/wordpress#82 Add support for creating WordPress accounts from the contact record

mlutfy opened the pull request:
https://lab.civicrm.org/dev/wordpress/-/issues/82

Overview
----------------------------------------

To reproduce:

- As a WP admin, create a new contact record that has an email
- Click on the Actions menu

Expected: an option to create a new CMS account for this contact.

References:

- https://civicrm.stackexchange.com/questions/9092/wordpress-creating-new-wp-users-from-civi-contacts

Before
----------------------------------------

Users cannot create WP accounts from CiviCRM.

After
----------------------------------------

Users can create WP accounts.

Technical Details / Comments
----------------------------------------

I wasn't too sure how to avoid the automatic login. It didn't seem very obvious how this is handled in Drupal/Joomla.

There is one WIP issue: after the account is created, we are redirected back to the front page of the site, instead of the contact record. Everything seems fine, except the redirect, which looks like an exception bounce?

mlutfy commented:
I tested:

* It doesn't seem to be throwing an exception (I added a try/catch block in the BAO)
* Tried adding `replaceUserContext` from the `createUser` function in WP.

christianwach commented on CRM/Utils/System/WordPress.php:
```diff
@@ -789,7 +789,12 @@ public function createUser(&$params, $mail) {
+    // Call wp_signon if we aren't already logged in
+    // For example, we might be creating a new user from the Contact record.
+    if (!current_user_can('create_users')) {
+      wp_signon($creds, FALSE);
```
I've always found this problematic but have never opened an issue for it...

It seems like bad practice to log a user in before confirming their email address in the normal WordPress fashion. The default email sent to the new User by `wp_new_user_notification()` below looks like this by default:

```
Username: <username>

To set your password, visit the following address:

https://domain.xyz/wp-login.php?action=rp&key=<key>&login=<username>

https://domain.xyz/wp-login.php
```

I would be in favour of removing it completely.

kcristiano replied:
I agree.  I am not seeing the need for this.  The comment mentions that we are not logged in, but to get to the create user account action I cannot think of an instance where we would not be logged in.

christianwach replied:
@kcristiano As discussed, `createUser()` is called from both CiviCRM Admin and front-end Profile Forms. In my opinion, it would be nice if Profile Forms had an option to enable/disable auto-login.

christianwach commented on CRM/Utils/System/WordPress.php:
```diff
@@ -990,6 +995,13 @@ public function alterAssetUrl(\Civi\Core\Event\GenericHookEvent $e) {
     }
   }
 
+  /**
+   * @inheritDoc
+   */
+  public function checkPermissionAddUser() {
+    return current_user_can('create_users');
+  }
+
```
This is all that's necessary for the "Create User Record" link to appear in the "Actions" menu.

christianwach commented:
> There is one WIP issue: after the account is created, we are redirected back to the front page of the site, instead of the contact record. Everything seems fine, except the redirect, which looks like an exception bounce?

Hmm, I do get redirected to the Contact record in CiviCRM 5.31. I wonder what's different in your install?

christianwach commented:
@mlutfy Other than the `wp_signon` issue, the ability to create a User from a Contact like this is neat 👍

mlutfy commented:
Thanks @christianwach !

Regarding `wp_signon`, I agree it's not the normal WP behaviour, however this would be a change to CiviCRM/WordPress that many people would consider a breaking change. I think it would require more discussion and should be clearly documented in its own PR.

(I had a quick look at Drupal, and from the cryptic code, it does not seem to login automatically. Although I have to admit it's pretty convenient to login automatically!)

christianwach commented:
> Regarding `wp_signon` [...] I think it would require more discussion and should be clearly documented in its own PR.

@mlutfy Fair enough - at least this PR prevents it being called unnecessarily. 🤷

kcristiano commented:
@mlutfy Testing on master

- If a contact does not have an email the option still appears.   However, we cannot add an email on the user creation screen.   Can we remove the option if the contact does not have an email?
- Can we remove the password fields and all WP to set the password?   I'd prefer the user set the password via the magic link in the email that the new user should get.
- I am not getting a confirmation email as the user, just the confirm to the administrator
- The bounce bothers me and I get it on master.  I'll test on 5.31 and see if we can see why there is a difference

mattwire commented:
> If a contact does not have an email the option still appears. However, we cannot add an email on the user creation screen. Can we remove the option if the contact does not have an email?

It would be nice to have a more obvious UI clue as to why we can't create a user. Maybe the option still there but pops up a dialog saying "you need to add an email to the contact first" or similar.

christianwach commented:
>> There is one WIP issue: after the account is created, we are redirected back to the front page of the site, instead of the contact record. Everything seems fine, except the redirect, which looks like an exception bounce?
>
> Hmm, I do get redirected to the Contact record in CiviCRM 5.31. I wonder what's different in your install?

@mlutfy Okay, so I've figured out why this is.

I have CiviCRM WP Profile Sync installed and this catches the problem. What's happening is that `createUser()` does not unhook the reverse sync that [takes place in the CiviCRM-WordPress plugin](https://github.com/civicrm/civicrm-wordpress/blob/145e98bb647122e66a6a624ed17f58846319e7e7/includes/civicrm.users.php#L85). So when the WordPress User is created in CiviCRM Admin, `CRM_Core_BAO_UFMatch::synchronizeUFMatch` is called which tries to set the session for the *created user* which causes CiviCRM to freak out and bounce you to the homepage.

The solution is to add methods [such as these](https://github.com/christianwach/civicrm-wp-profile-sync/blob/02954d303ee9c1b5843427f7fe55b91759232b23/civicrm-wp-profile-sync.php#L1041-L1118) to the `CRM_Utils_System_WordPress` class, then replace:

```php
$uid = wp_insert_user($user_data);
```

with

```php
$this->hooks_core_remove();
$uid = wp_insert_user($user_data);
$this->hooks_core_add();
```

FWIW, these unhook/re-hook calls should take place any time CiviCRM Core updates the WordPress User.

christianwach commented on CRM/Utils/System/WordPress.php:
```diff
@@ -789,7 +789,12 @@ public function createUser(&$params, $mail) {
+      wp_signon($creds, FALSE);
+    }
 
     wp_new_user_notification($uid, $user_data['user_pass']);
```
@mlutfy One more thing while this is getting attention - this call to `wp_new_user_notification()` needs updating. The [docs say](https://developer.wordpress.org/reference/functions/wp_new_user_notification/) that the second param has been deprecated since WordPress 4.3. The current params are:

```php
wp_new_user_notification( int $user_id, null $deprecated = null, string $notify = '' )
```

If both admin(s) _and_ the user need to be notified, then (because admin-only is the default) it should read:

```php
wp_new_user_notification($uid, null, 'both');
```

mlutfy commented:
Thanks everyone for the really great feedback.

I fixed the following issues:

* Bouncing to the front page: I copied Christian's code to remove/add the user hooks.
  * had to re-add the hooks after notifying to avoid the bouncing
  * removed references to the "old-style plugin structure" - was this for old versions of CiviCRM?
  * made the function public, in case plugins want to use them.
* Removed the password field. This also applies to public profiles.
* Updated the `wp_new_user_notification` call.

Not fixed:

* When a contact record does not have an email, the feedback is initially not clear. If we submit the form, we do see a "email is required" error:

![civi-wp-cms2020-11-21_15-43](https://user-images.githubusercontent.com/254741/99887731-c91cad80-2c14-11eb-982c-cef8db14aa50.png)

The email field is frozen, so presumably this was an afterthought hack. I think having a bounce might be better (with an error message), but if there are no strong feelings about it, I would rather punt it, since it's a more general UX issue that affects all CMSes.

christianwach commented:
@mlutfy Looks like this is getting there 👍  You can remove the two calls to `do_action('civicrm_wp_profile_sync_hooks_core_added');` as they're specific to my plugin.

christianwach commented:
> removed references to the "old-style plugin structure" - was this for old versions of CiviCRM?

@mlutfy Yes. Separate plugins can be active with any version of CiviCRM. Not needed here.

mlutfy commented:
Perfect, thanks for the quick review. I updated my PR to remove the profile-sync-specific actions.

christianwach commented:
@mlutfy Related ticket: https://lab.civicrm.org/dev/wordpress/-/issues/83

I have asked Bradley to re-post his thoughts here for your attention.

christianwach commented on CRM/Utils/System/WordPress.php:
```diff
@@ -783,15 +782,22 @@ public function createUser(&$params, $mail) {
+      wp_signon($creds, FALSE);
+    }
+
+    wp_new_user_notification($uid, NULL, 'both');
```
@mlutfy FWIW I agree with this proposal. Using `do_action('register_new_user')` instead of calling `wp_new_user_notification()` directly seems like the correct way of doing this.

mlutfy commented:
hi @braders, thank you for the feedback, and sorry for the slow response.

I updated my PR, but I have to admit I am not very familiar with WP internals, could you double-check if I got it right?

i.e.

```
-    wp_new_user_notification($uid, NULL, 'both');
+   do_action('register_new_user', $uid)
```

braders commented:
> hi @braders, thank you for the feedback, and sorry for the slow response.
> 
> I updated my PR, but I have to admit I am not very familiar with WP internals, could you double-check if I got it right?
> 
> i.e.
> 
> ```
> -    wp_new_user_notification($uid, NULL, 'both');
> +   do_action('register_new_user', $uid);
> ```

Looks good to me!

kcristiano commented:
@eileenmcnaughton i have done testing with this and would like to see it merged.   

I've done `r-run` in all the use cases we have and it passes.   The new functionality is a great improvement and works as described.

The earlier we can get merged the better so we can get it eyes on it before it releases as it will be updating how users are created via profile to the new standards in WP.  - No more users setting passwords - magic links!

mlutfy commented:
@kcristiano @braders @christianwach  Thanks for the reviews! I took the liberty to merge based on your reviews, and the rapidly approaching branch date for 5.37. I've been running this in production and so far so good.

eileenmcnaughton commented:
I agree that @kcristiano's comment was all that merge was pending
